File: lib/structs.go
package lib

// Task represents a task that is assigned to NMS_Agents.
type Task struct {
	Task_ID   string   `json:"task_id"`   // Unique id task
	Frequency int      `json:"frequency"` // Frequency (seconds)
	Devices   []Device `json:"devices"`   // List of devices included in the task
}

// Device represents a network device (router for example) and the metrics that are to be monitored.
type Device struct {
	Device_ID      string         `json:"device_id"`      // Unique id device
	Device_Metrics Device_Metrics `json:"device_metrics"` // Metrics of the device
	Link_Metrics   Link_Metrics   `json:"link_metrics"`   // Network link metrics for inter-device communication
}

// Device_Metrics holds resource information for a network device.
type Device_Metrics struct {
	CPU_Usage       bool     `json:"cpu_usage"`       // Indicates if CPU is active
	RAM_Usage       bool     `json:"ram_usage"`       // Indicates if RAM is active
	Interface_Stats []string `json:"interface_stats"` // List of network interface names to monitor
}

// Link_Metrics represents network metrics for communication between devices.
type Link_Metrics struct {
	Bandwidth            Metric_Test_Config `json:"bandwidth"`            // Configuration for bandwidth testing
	Jitter               Metric_Test_Config `json:"jitter"`               // Configuration for jitter measurement
	Packet_Loss          Metric_Test_Config `json:"packet_loss"`          // Configuration for packet loss measurement
	Latency              Metric_Test_Config `json:"latency"`              // Configuration for latency measurement
	AlertFlow_Conditions Alert_Conditions   `json:"alertflow_conditions"` // Alert conditions for each metric
}

// Metric_Test_Config represents the configuration settings for network tests.
type Metric_Test_Config struct {
	Tool           string `json:"tool"`           // Name of the tool used for the test (iperf, etc.)
	Is_Client      bool   `json:"is_client"`      // True if the device is a client in this test
	Server_Address string `json:"server_address"` // Server address for the test
	Duration       int    `json:"duration"`       // Duration of the test (seconds)
	Transport_Type string `json:"transport_type"` // Type of transport protocol
	Frequency      int    `json:"frequency"`      // Frequency of the test (seconds)
}

// Alert_Conditions defines the limit for triggering alerts based on specific metrics.
// If any metric exceeds its limit, the NMS_Agent will notify the NMS_Server.
type Alert_Conditions struct {
	CPU_Usage       int `json:"cpu_usage"`       // CPU usage limit (%)
	RAM_Usage       int `json:"ram_usage"`       // RAM usage limit (%)
	Interface_Stats int `json:"interface_stats"` // Limit for packets per second on interfaces
	Packet_Loss     int `json:"packet_loss"`     // Packet loss limit (%)
	Jitter          int `json:"jitter"`          // Jitter limit (milliseconds)
}

File: src/lib/structs.py
class Task:
    def __init__(self, taskR):
        self.task_id = taskR.get("task_id")
        self.task_frequency = taskR.get("frequency")
        self.devices = [Device(deviceR["device_id"], deviceR["device_metrics"], deviceR["link_metrics"]) for deviceR in taskR.get("devices", [])]

    def show(self):
        print(f"Task:\n Task_ID: {self.task_id} \n frequecy: {self.task_frequency}")
        for device in self.devices:
            device.show()

class Device:
    def __init__(self, device_id, device_metrics, link_metrics):
        self.device_id = device_id
        if isinstance(device_metrics, dict):
            self.device_metrics = DeviceMetrics(**device_metrics)  # Converte dicionário em objeto
        elif isinstance(device_metrics, DeviceMetrics):
            self.device_metrics = device_metrics  # Já é um objeto, então usa diretamente
        else:
            raise ValueError("device_metrics deve ser um dicionário ou um objeto DeviceMetrics")
        
        # Verifica se link_metrics é um dicionário ou já é um objeto LinkMetrics
        if isinstance(link_metrics, dict):
            self.link_metrics = LinkMetrics(**link_metrics)  # Converte dicionário em objeto
        elif isinstance(link_metrics, LinkMetrics):
            self.link_metrics = link_metrics  # Já é um objeto, então usa diretamente
        else:
            raise ValueError("link_metrics deve ser um dicionário ou um objeto LinkMetrics")
    
    
    def show(self):
        print(f"  Device_ID: {self.device_id}")
        print("  Device_metrics:")
        self.device_metrics.show()
        print("  Link_metrics:")
        self.link_metrics.show()

class DeviceMetrics:
    def __init__(self, cpu_usage, ram_usage, interface_stats):
        self.cpu_usage = cpu_usage
        self.ram_usage = ram_usage
        self.interface_stats = interface_stats

    def show(self):
        print(f"    CPU_usage: {self.cpu_usage}")
        print(f"    RAM_usage: {self.ram_usage}")
        print(f"    Interface_stats: {', '.join(self.interface_stats)}")

class LinkMetrics:
    def __init__(self, bandwidth, jitter, packet_loss, latency, alertflow_conditions):
        self.bandwidth = BandwidthMetric(**bandwidth)
        self.jitter = JitterMetric(**jitter)
        self.packet_loss = PacketLossMetric(**packet_loss)
        self.latency = LatencyMetric(**latency)
        self.alertflow_conditions = AlertFlowConditions(**alertflow_conditions)

    def show(self):
        print("    Bandwith:")
        self.bandwidth.show()
        print("    Jitter:")
        self.jitter.show()
        print("   Packet_loss:")
        self.packet_loss.show()
        print("    Latency:")
        self.latency.show()
        print("    AlertFlow_conditions:")
        self.alertflow_conditions.show()

class BandwidthMetric:
    def __init__(self, tool, is_server, server_address, duration, transport, frequency):
        self.tool = tool
        self.is_server = is_server
        self.server_address = server_address
        self.duration = duration
        self.transport = transport
        self.frequency = frequency

    def show(self):
        print(f"      Tool: {self.tool}, Is_Server: {self.is_server}, Server_Address: {self.server_address}")
        print(f"      Duration: {self.duration}, Transport: {self.transport}, Frequency: {self.frequency}")

class JitterMetric():
    def __init__(self, tool, is_server, server_address, duration, transport, frequency):
        self.tool = tool
        self.is_server = is_server
        self.server_address = server_address
        self.duration = duration
        self.transport = transport
        self.frequency = frequency

    def show(self):
        print(f"      Tool: {self.tool}, Is_Server: {self.is_server}, Server_Address: {self.server_address}")
        print(f"      Duration: {self.duration}, Transport: {self.transport}, Frequency: {self.frequency}")

class PacketLossMetric():
    def __init__(self, tool, is_server, server_address, duration, transport, frequency):
        self.tool = tool
        self.is_server = is_server
        self.server_address = server_address
        self.duration = duration
        self.transport = transport
        self.frequency = frequency

    def show(self):
        print(f"      Tool: {self.tool}, Is_Server: {self.is_server}, Server_Address: {self.server_address}")
        print(f"      Duration: {self.duration}, Transport: {self.transport}, Frequency: {self.frequency}")

class LatencyMetric:
    def __init__(self, tool, destination_address, packet_count, frequency):
        self.tool = tool
        self.destination_address = destination_address
        self.packet_count = packet_count
        self.frequency = frequency

    def show(self):
        print(f"      Tool: {self.tool}, Destination_address: {self.destination_address}")
        print(f"      Packet_count: {self.packet_count}, Frequency: {self.frequency}")

class AlertFlowConditions:
    def __init__(self, cpu_usage, ram_usage, interface_stats, packet_loss, jitter):
        self.cpu_usage = cpu_usage
        self.ram_usage = ram_usage
        self.interface_stats = interface_stats
        self.packet_loss = packet_loss
        self.jitter = jitter

    def show(self):
        print(f"      CPU_usage: {self.cpu_usage}, RAM_usage: {self.ram_usage}")
        print(f"      Interface_Stats: {self.interface_stats}, Packet_loss: {self.packet_loss}")
        print(f"      Jitter: {self.jitter}")
